src/prop/sat_solver_factory.h

```diff
@@ -29,29 +30,62 @@
 namespace cvc5::internal {
 namespace prop {
 
+class TheoryProxy;
+
 class SatSolverFactory
 {
  public:
-  static CDCLTSatSolver* createCDCLTMinisat(Env& env,
-                                            StatisticsRegistry& registry);
-
-  static CDCLTSatSolver* createCadical(Env& env,
-                                       StatisticsRegistry& registry,
-                                       ResourceManager* resmgr,
-                                       const std::string& name = "");
-
-  static CDCLTSatSolver* createCadicalCDCLT(Env& env,
-                                            StatisticsRegistry& registry,
-                                            ResourceManager* resmgr,
-                                            const std::string& name = "");
-
-  static SatSolver* createCryptoMinisat(StatisticsRegistry& registry,
-                                        ResourceManager* resmgr,
-                                        const std::string& name = "");
-
-  static SatSolver* createKissat(StatisticsRegistry& registry,
-                                 const std::string& name = "");
-}; /* class SatSolverFactory */
+  enum SatSolverType
+  {
+    MINISAT,
+    CADICAL,
+    CRYPTOMINISAT,
+    KISSAT,
+  };
+
+  template <SatSolverType T>
+  static SatSolver* createSatSolver(Env& env,
+                                    StatisticsRegistry& registry,
+                                    ResourceManager* resmgr,
+                                    const std::string& name = "");
+
+  template <SatSolverType T>
+  static CDCLTSatSolver* createCDCLTSatSolver(Env& env,
+                                              StatisticsRegistry& registry,
+                                              ResourceManager* resmgr,
+                                              TheoryProxy* theory_proxy,
+                                              const std::string& name = "");
+};
+
+template <>
+SatSolver* SatSolverFactory::createSatSolver<SatSolverFactory::CADICAL>(
+    Env&, StatisticsRegistry&, ResourceManager*, const std::string&);
+
+template <>
+SatSolver* SatSolverFactory::createSatSolver<SatSolverFactory::KISSAT>(
+    Env&, StatisticsRegistry&, ResourceManager*, const std::string&);
+
+template <>
+SatSolver* SatSolverFactory::createSatSolver<SatSolverFactory::CRYPTOMINISAT>(
+    Env&, StatisticsRegistry&, ResourceManager*, const std::string&);
+
+template <>
+CDCLTSatSolver*
+SatSolverFactory::createCDCLTSatSolver<SatSolverFactory::MINISAT>(
+    Env&,
+    StatisticsRegistry&,
+    ResourceManager*,
+    TheoryProxy*,
+    const std::string&);
+
+template <>
+CDCLTSatSolver*
+SatSolverFactory::createCDCLTSatSolver<SatSolverFactory::CADICAL>(
+    Env&,
+    StatisticsRegistry&,
+    ResourceManager*,
+    TheoryProxy*,
+    const std::string&);
```

Comment:
I'm not sure this refactor has any advantage over two methods createSatSolver/createCDCLTSatSolver that take the `BvSatSolverMode/SatSolverMode` option enums to create an instance of the given solver. We already maintain two enums for BV/CDCLT sat solvers, we shouldn't need a third one.

Reply:
The code is, in a sense cleaner as adding/removing a solver does not need to touch any code of any other solver. It also reflects the previous version where there was one function for each solver (without the disadvantage of diverging function prototypes).
And it allows to generate a function pointer to a solver generator, which is not possible with the type of solver is a parameter.
Using the existing enums is a good idea, though.